File: personal-assistant-agent/src/tools/FetchDailyMeetingSchedule.ts
import { google } from 'googleapis';
import type { OAuth2Client } from 'google-auth-library';
import { localDayBounds, localTimeZone } from '../utils/dateUtils';

export interface CalendarEventSummary {
    summary: string;
    timeWindow: string;
    link?: string | null;
}

export interface DailyScheduleResult {
    date: string;
    timeZone: string;
    events: CalendarEventSummary[];
}

function formatEventTime(isoOrDate: string): string {
    if (!isoOrDate.includes('T')) {
        return 'All day';
    }
    const date = new Date(isoOrDate);
    return date.toLocaleTimeString('en-US', {
        hour: 'numeric',
        minute: '2-digit',
        hour12: true,
    });
}

export const fetchDailyMeetingSchedule = async (
    auth: OAuth2Client,
    date: string
): Promise<DailyScheduleResult> => {
    const calendar = google.calendar({ version: 'v3', auth });
    const timeZone = localTimeZone();
    const { timeMin, timeMax } = localDayBounds(date);

    const response = await calendar.events.list({
        calendarId: 'primary',
        timeMin,
        timeMax,
        timeZone,
        singleEvents: true,
        orderBy: 'startTime',
    });

    const events = response.data.items || [];

    const mapped: CalendarEventSummary[] = events.map((event) => {
        const startRaw = event.start?.dateTime || event.start?.date || '';
        const endRaw = event.end?.dateTime || event.end?.date || '';

        const startLabel = formatEventTime(startRaw);
        const endLabel = endRaw.includes('T') ? formatEventTime(endRaw) : '';

        return {
            summary: event.summary || 'Untitled Event',
            timeWindow: endLabel ? `${startLabel} – ${endLabel}` : startLabel,
            link: event.htmlLink,
        };
    });

    return {
        date,
        timeZone,
        events: mapped,
    };
};

File: personal-assistant-agent/src/tools/fetch_daily_meeting_schedule.py
from dataclasses import dataclass
from datetime import datetime
from typing import Any, List, Optional

from google.oauth2.credentials import Credentials
from googleapiclient.discovery import build

from src.utils.date_utils import local_day_bounds, local_time_zone


@dataclass
class CalendarEventSummary:
    summary: str
    time_window: str
    link: Optional[str] = None


@dataclass
class DailyScheduleResult:
    date: str
    time_zone: str
    events: List[CalendarEventSummary]

    def to_dict(self) -> dict[str, Any]:
        return {
            "date": self.date,
            "timeZone": self.time_zone,
            "events": [
                {
                    "summary": e.summary,
                    "timeWindow": e.time_window,
                    "link": e.link,
                }
                for e in self.events
            ],
        }


def _format_event_time(iso_or_date: str) -> str:
    if "T" not in iso_or_date:
        return "All day"
    normalized = iso_or_date.replace("Z", "+00:00")
    dt = datetime.fromisoformat(normalized)
    if dt.tzinfo:
        dt = dt.astimezone()
    return dt.strftime("%I:%M %p").lstrip("0") or dt.strftime("%I:%M %p")


def fetch_daily_meeting_schedule(
    auth: Credentials, date: str
) -> DailyScheduleResult:
    calendar = build("calendar", "v3", credentials=auth)
    time_zone = local_time_zone()
    bounds = local_day_bounds(date)

    response = (
        calendar.events()
        .list(
            calendarId="primary",
            timeMin=bounds["timeMin"],
            timeMax=bounds["timeMax"],
            timeZone=time_zone,
            singleEvents=True,
            orderBy="startTime",
        )
        .execute()
    )

    events = response.get("items") or []
    mapped: List[CalendarEventSummary] = []

    for event in events:
        start_raw = (event.get("start") or {}).get("dateTime") or (event.get("start") or {}).get(
            "date"
        ) or ""
        end_raw = (event.get("end") or {}).get("dateTime") or (event.get("end") or {}).get(
            "date"
        ) or ""

        start_label = _format_event_time(start_raw)
        end_label = _format_event_time(end_raw) if "T" in end_raw else ""

        mapped.append(
            CalendarEventSummary(
                summary=event.get("summary") or "Untitled Event",
                time_window=f"{start_label} – {end_label}" if end_label else start_label,
                link=event.get("htmlLink"),
            )
        )

    return DailyScheduleResult(date=date, time_zone=time_zone, events=mapped)
